modules/programs/neovim/default.nix

```
@@ -0,0 +1,96 @@
+{ mode, config, pkgs, lib, soxin, ... }:
+
+with lib;
+let
+  cfg = config.soxin.programs.neovim;
+in
+{
+  options = {
+    soxin.programs.neovim = soxin.lib.mkSoxinModule {
+      inherit config;
+      name = "neovim";
+      includeKeyboardLayout = true;
+      includeTheme = true;
+      extraOptions = {
+        extraConfig = mkOption {
+          type = types.lines;
+          default = "";
+          example = ''
+            set nocompatible
+            set nobackup
+          '';
+          description = ''
+            Custom vimrc lines.
+          '';
+        };
+
+        plugins = mkOption {
+          type = with types; listOf (either package soxin.lib.modules.neovim.pluginWithConfigModule);
+          default = [ ];
+          defaultText = "The plugins added by the theme";
+          example = literalExample ''
+            with pkgs.vimPlugins; [
+              yankring
+              vim-nix
+              { plugin = vim-startify;
+                config = "let g:startify_change_to_vcs_root = 0";
+              }
+            ]
+          '';
+          description = ''
+            List of vim plugins to install optionally associated with
+            configuration to be placed in init.vim.
+          '';
+        };
+      };
+    };
+  };
+
+  config = mkIf cfg.enable (mkMerge [
+    # add all plugins installed by themes
+    { soxin.programs.neovim.plugins = cfg.theme.plugins; }
+
+    (optionalAttrs (mode == "NixOS") {
+      # TODO: Add support for NeoVim on NixOS.
+      # programs.neovim = {
+      #   enable = true;
+      #
+      #   # make it a default editor
+      #   defaultEditor = true;
+      #
+      #   # Create a symlink of nvim to vi and vim
+      #   viAlias = true;
+      #   vimAlias = true;
+      #
+      #   # Add support for NodeJS, Python 2 and 3 as well as Ruby
+      #   # withNodeJs = true;
+      #   # withPython3 = true;
+      #   withRuby = true;
+      #
+      #   # Add the Python's neovim plugin
+      #   # extraPython3Packages = ps: with ps; [ pynvim ];
+      #
+      #   configure = {
+      #     customRC = extraRC;
+      #     packages.myVimPackage = { inherit start opt; };
+      #   };
+      # };
+    })
+
+    (optionalAttrs (mode == "home-manager") {
+      # TODO: I wish home-manager had a defaultEditor as well!
+      home.sessionVariables = { EDITOR = "nvim"; };
+
+      programs.neovim = {
+        inherit (cfg) enable extraConfig plugins;
```

```suggestion
        enable = true;
        inherit (cfg) extraConfig plugins;
```
We have an `mkIf` above.

---

Very true, it was just easier to type an inherit for all fields. Is it confusing enough to merit creating two lines for it?

---

It isn't especially confusing, it's only for consistency.